From 6a5f5fec50162f56555874f46334672e47a3c62b Mon Sep 17 00:00:00 2001
From: Andi Kleen <ak@linux.intel.com>
Date: Sat, 3 Apr 2021 22:23:56 -0700
Subject: [PATCH 08/39] static_call: Use single copy of static_call_return0

With the inline version of static calls it is trying to use
all static call functions inline. But this doesn't work for
__static_call_return0, because its address is always taken,
which forces the compiler to generate a out of line copy.

If it only exists as a static inline this means there are
many copies generated. Instead use the out of line in static_call.c
for this.

This fixes another bug. When _INLINE is set static_inline.c was
not compiled at all, which disabled the self test even when
it was enabled in the configuration.

This fixes a build problem with gcc LTO. __static_call_return0
is referenced from assembler, which requires making it global
because the assembler can end in a different file than the other
C code.  But that's not possible for a static inline function.

Signed-off-by: Andi Kleen <ak@linux.intel.com>
---
 include/linux/static_call.h | 9 ++-------
 kernel/Makefile             | 2 +-
 kernel/static_call.c        | 4 ++++
 3 files changed, 7 insertions(+), 8 deletions(-)

diff --git a/include/linux/static_call.h b/include/linux/static_call.h
index ae04b1123a93..e75aef659224 100644
--- a/include/linux/static_call.h
+++ b/include/linux/static_call.h
@@ -100,6 +100,8 @@
 #ifdef CONFIG_HAVE_STATIC_CALL
 #include <asm/static_call.h>
 
+extern long __static_call_return0(void);
+
 /*
  * Either @site or @tramp can be NULL.
  */
@@ -148,8 +150,6 @@ extern void __static_call_update(struct static_call_key *key, void *tramp, void
 extern int static_call_mod_init(struct module *mod);
 extern int static_call_text_reserved(void *start, void *end);
 
-extern long __static_call_return0(void);
-
 #define __DEFINE_STATIC_CALL(name, _func, _func_init)			\
 	DECLARE_STATIC_CALL(name, _func);				\
 	__visible struct static_call_key STATIC_CALL_KEY(name) = {	\
@@ -221,11 +221,6 @@ static inline int static_call_text_reserved(void *start, void *end)
 	return 0;
 }
 
-static inline long __static_call_return0(void)
-{
-	return 0;
-}
-
 #define EXPORT_STATIC_CALL(name)					\
 	EXPORT_SYMBOL(STATIC_CALL_KEY(name));				\
 	EXPORT_SYMBOL(STATIC_CALL_TRAMP(name))
diff --git a/kernel/Makefile b/kernel/Makefile
index b6407f92c9af..aaa1dae69352 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -111,7 +111,7 @@ obj-$(CONFIG_CPU_PM) += cpu_pm.o
 obj-$(CONFIG_BPF) += bpf/
 obj-$(CONFIG_KCSAN) += kcsan/
 obj-$(CONFIG_SHADOW_CALL_STACK) += scs.o
-obj-$(CONFIG_HAVE_STATIC_CALL_INLINE) += static_call.o
+obj-$(CONFIG_HAVE_STATIC_CALL) += static_call.o
 
 obj-$(CONFIG_PERF_EVENTS) += events/
 
diff --git a/kernel/static_call.c b/kernel/static_call.c
index 6d332c0c9134..7ec8608d94ab 100644
--- a/kernel/static_call.c
+++ b/kernel/static_call.c
@@ -10,6 +10,8 @@
 #include <linux/processor.h>
 #include <asm/sections.h>
 
+#ifdef CONFIG_HAVE_STATIC_CALL_INLINE
+
 extern struct static_call_site __start_static_call_sites[],
 			       __stop_static_call_sites[];
 extern struct static_call_tramp_key __start_static_call_tramp_key[],
@@ -496,6 +498,8 @@ int __init static_call_init(void)
 }
 early_initcall(static_call_init);
 
+#endif /* HAVE_STATIC_CALL_INLINE */
+
 long __static_call_return0(void)
 {
 	return 0;
-- 
2.30.2

